devel/libert_util/src/perm_vector.c

```diff
@@ -0,0 +1,76 @@
+/*
+   Copyright (C) 2016  Statoil ASA, Norway.
+
+   The file 'perm_vector.c' is part of ERT - Ensemble based Reservoir Tool.
+
+   ERT is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   ERT is distributed in the hope that it will be useful, but WITHOUT ANY
+   WARRANTY; without even the implied warranty of MERCHANTABILITY or
+   FITNESS FOR A PARTICULAR PURPOSE.
+
+   See the GNU General Public License at <http://www.gnu.org/licenses/gpl.html>
+   for more details.
+*/
+#include <stdlib.h>
+
+#include <ert/util/type_macros.h>
+#include <ert/util/util.h>
+#include <ert/util/perm_vector.h>
+
+#define PERM_VECTOR_TYPE_ID 661433
+
+struct perm_vector_struct {
+  UTIL_TYPE_ID_DECLARATION;
+  int   size;
+  int * perm;
+};
+
+
+
+perm_vector_type * perm_vector_alloc( int size ) {
+  perm_vector_type * perm = util_malloc( sizeof * perm );
+  UTIL_TYPE_ID_INIT( perm , PERM_VECTOR_TYPE_ID );
+  perm->size = size;
+  perm->perm = util_malloc( size * sizeof * perm->perm );
+  {
+    int i;
+    for (i=0; i<size; i++)
+      perm->perm[i] = i;
+  }
+
+  return perm;
+}
+
+
+void perm_vector_free( perm_vector_type * perm) {
+  free( perm->perm );
+  free( perm );
+}
+
+
+
+int perm_vector_get_size( const perm_vector_type * perm) {
+  return perm->size;
+}
+
+
+int perm_vector_iget( const perm_vector_type * perm, int index) {
+  if (index < perm->size)
+    return perm->perm[index];
+  else {
+    util_abort("%s: invalid index:%d \n",__func__ , index );
+    return -1;
+  }
+}
+
+
+void perm_vector_iset( perm_vector_type * perm, int index, int value) {
```

Comment:
> but isn't it somewhat "permissive"? What I would expect would be that you have to take two indices that you then flip

Yes - it is quite permissive; I tried to implement it as a swap with two indices - but those feeble attempts were quite far from O(1). This is the reason there is no `__setitem__` implemented in Python - I will at least add a comment.